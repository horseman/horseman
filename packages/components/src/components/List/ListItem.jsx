import styled, { css } from "styled-components";
import PropTypes from "prop-types";

import media from "../../styleUtils/media";

/**
 * Calculate the width based on the mobile base and the gutter
 */
const getItemWidth = (mobileBase, gutter) => {
  if (mobileBase > 1) {
    return `calc(${100 / mobileBase}% - ${gutter / 2}px)`;
  }
  return `${100 / mobileBase}%`;
};

const ListItem = styled.div`
  width: ${({ mobileBase, gutter }) => getItemWidth(mobileBase, gutter)};
  margin-top: ${({ gutter }) => gutter}px;

  &:nth-child(-n + ${props => props.mobileBase}) {
    margin-top: 0;
  }

  ${({ breakpoint }) =>
    media[breakpoint](css`
      width: calc(${({ base, gutter }) => `${100 / base}% - ${gutter / 2}px`});
      &:nth-child(-n + ${({ base }) => base}) {
        margin-top: 0em;
      }
    `)};
`;

ListItem.defaultProps = {
  gutter: 30,
  mobileBase: 1,
  breakpoint: "mdAndUp",
};

ListItem.propTypes = {
  base: PropTypes.number.isRequired,
  mobileBase: PropTypes.number,

  /**
   * The breakpoint at which the layout will go desktop to mobile
   */
  breakpoint: PropTypes.string,
};

export default ListItem;
