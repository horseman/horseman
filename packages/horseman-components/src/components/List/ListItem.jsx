import styled, { css } from "styled-components";
import PropTypes from "prop-types";

import media from "../../styleUtils/media";

/**
 * Calculate the width based on the mobile base and the gutter
 */
const getItemWidth = (mobileBase, gutter) => {
  if (mobileBase > 1) {
    return `calc(${100 / mobileBase}% - ${gutter / 2}px)`;
  }
  return `${100 / mobileBase}%`;
};

const ListItem = styled.div`
  width: ${({ mobileBase, gutter }) => getItemWidth(mobileBase, gutter)};
  margin-top: ${({ gutter }) => gutter}px;

  &:nth-child(-n + ${props => props.mobileBase}) {
    margin-top: 0;
  }

  ${media.mdAndUp(css`
    width: calc(${({ base, gutter }) => `${100 / base}% - ${gutter / 2}px`});
    &:nth-child(-n + ${({ base }) => base}) {
      margin-top: 0em;
    }
  `)};
`;

ListItem.defaultProps = {
  gutter: 30,
  mobileBase: 1,
};

ListItem.propTypes = {
  base: PropTypes.number.isRequired,
  mobileBase: PropTypes.number,
};

export default ListItem;
