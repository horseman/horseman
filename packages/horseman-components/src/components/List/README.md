A list is a structural component intended to render a list of items. You can
supply a mobile and desktop base that provide the number of items that will be
spread across a single line on mobile and desktop respectively.

The list will never allow an item to exist on a line by itself, so it will
always ensure that multiple items are on a line.

### Examples
```
<List base={5} mobileBase={2}>
  <div style={{background: 'gray', height: '100px'}}></div>
  <div style={{background: 'gray', height: '100px'}}></div>
  <div style={{background: 'gray', height: '100px'}}></div>
  <div style={{background: 'gray', height: '100px'}}></div>
  <div style={{background: 'gray', height: '100px'}}></div>
  <div style={{background: 'gray', height: '100px'}}></div>
</List>
```
