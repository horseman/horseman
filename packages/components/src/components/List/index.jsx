import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import ListItem from "./ListItem";

const StyledList = styled.div`
  display: flex;
  flex-wrap: wrap;
  flex-shrink: 0;
  flex-grow: 0;
  justify-content: space-between;
  align-items: bottom;
`;

const List = ({ children, base, mobileBase, breakpoint, rebase }) => {
  const numItems = children.length;

  // The list basis is the number of items that will live on a single line in
  // desktop. The number of items composing the listBasis will never be larger
  // than the total number of items.
  const listBasis = base > numItems ? numItems : base;

  // If the base we are given forces the list to have a trailing single item
  // we need to rebase in order to never leave a trailing item.
  const nonTrailingBase =
    numItems % listBasis === 1 ? listBasis - 1 : listBasis;

  // determine if rebase is desired, use appropriate base.
  const activeBase = rebase ? nonTrailingBase : listBasis;

  const trailingItems = numItems % activeBase;
  const emptySlots = trailingItems ? activeBase - trailingItems : 0;

  /* eslint-disable react/no-array-index-key */
  const items = children.map((child, i) => (
    <ListItem
      key={i}
      mobileBase={mobileBase}
      base={activeBase}
      breakpoint={breakpoint}
    >
      {child}
    </ListItem>
  ));

  const placeholders = Array(emptySlots)
    .fill()
    .map((_, i) => (
      <ListItem key={i} mobileBase={mobileBase} base={activeBase} />
    ));
  /* eslint-enable react/no-array-index-key */

  return (
    <StyledList>
      {items}
      {placeholders}
    </StyledList>
  );
};

List.defaultProps = {
  mobileBase: 1,
  rebase: true,
};

List.propTypes = {
  /**
   * The items that will be listed over.
   */
  children: PropTypes.node.isRequired,

  /**
   * This is the number of items that will be displayed across on desktop
   */
  base: PropTypes.oneOf([1, 2, 3, 4, 5, 6]).isRequired,

  /**
   * This is the number of items that will be displayed across on mobile
   */
  mobileBase: PropTypes.oneOf([1, 2, 3]).isRequired,

  /**
   * The breakpoint at which the layout will go desktop to mobile
   */
  breakpoint: PropTypes.string,

  /**
   * Will re-base the list to never allow a single item on a line. If set to
   * false will use `base` as it is passed in, otherwise will alter the active
   * base depending on the amount of ListItems present
   */
  rebase: PropTypes.bool,
};

/**
 * @component
 */
export default List;
