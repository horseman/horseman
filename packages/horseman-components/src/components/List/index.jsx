import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import ListItem from "./ListItem";

const StyledList = styled.div`
  display: flex;
  flex-wrap: wrap;
  flex-shrink: 0;
  flex-grow: 0;
  justify-content: space-between;
  align-items: bottom;
`;

const List = ({ children, base, mobileBase }) => {
  const numItems = children.length;

  // The list basis is the number of items that will live on a single line in
  // desktop. The number of items composing the listBasis will never be larger
  // than the total number of items.
  const listBasis = base > numItems ? numItems : base;

  // If the base we are given forces the list to have a trailing single item
  // we need to reduce the base in order to never leave a trailing item.
  const nonTrailingBase =
    numItems % listBasis === 1 ? listBasis - 1 : listBasis;

  const trailingItems = numItems % nonTrailingBase;
  const emptySlots = trailingItems ? nonTrailingBase - trailingItems : 0;

  /* eslint-disable react/no-array-index-key */
  const items = children.map((child, i) => (
    <ListItem key={i} mobileBase={mobileBase} base={nonTrailingBase}>
      {child}
    </ListItem>
  ));

  const placeholders = Array(emptySlots)
    .fill()
    .map((_, i) => (
      <ListItem key={i} mobileBase={mobileBase} base={nonTrailingBase} />
    ));
  /* eslint-enable react/no-array-index-key */

  return (
    <StyledList>
      {items}
      {placeholders}
    </StyledList>
  );
};

List.defaultProps = {
  mobileBase: 1,
};

List.propTypes = {
  /**
   * The items that will be listed over.
   */
  children: PropTypes.node.isRequired,

  /**
   * This is the number of items that will be displayed across on desktop
   */
  base: PropTypes.oneOf([1, 2, 3, 4, 5, 6]).isRequired,

  /**
   * This is the number of items that will be displayed across on mobile
   */
  mobileBase: PropTypes.oneOf([1, 2, 3]).isRequired,
};

/**
 * @component
 */
export default List;
