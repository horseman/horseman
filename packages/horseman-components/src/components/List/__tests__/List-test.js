import React from "react";
import { shallow } from "enzyme";

import List from "../";

describe("List", () => {
  test("renders with a base and default mobileBase", () => {
    const wrapper = shallow(
      <List base={2}>
        <div />
        <div />
      </List>,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  test("modifies the basis to prevent trailing items.", () => {
    const wrapper = shallow(
      <List base={3}>
        <div />
        <div />
        <div />
        <div />
      </List>,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  test("Create the correct number of trailing empty children to force display", () => {
    const wrapper = shallow(
      <List base={3}>
        <div />
        <div />
        <div />
        <div />
        <div />
      </List>,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});
