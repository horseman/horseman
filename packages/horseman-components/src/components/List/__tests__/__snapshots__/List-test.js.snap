// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Create the correct number of trailing empty children to force display 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-shrink: 0;
  -ms-flex-shrink: 0;
  flex-shrink: 0;
  -webkit-flex-grow: 0;
  -ms-flex-grow: 0;
  flex-grow: 0;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: bottom;
  -webkit-box-align: bottom;
  -ms-flex-align: bottom;
  align-items: bottom;
}

<div
  className="c0"
>
  <styled.div
    base={3}
    breakpoint="mdAndUp"
    gutter={30}
    key="0"
    mobileBase={1}
  >
    <div />
  </styled.div>
  <styled.div
    base={3}
    breakpoint="mdAndUp"
    gutter={30}
    key="1"
    mobileBase={1}
  >
    <div />
  </styled.div>
  <styled.div
    base={3}
    breakpoint="mdAndUp"
    gutter={30}
    key="2"
    mobileBase={1}
  >
    <div />
  </styled.div>
  <styled.div
    base={3}
    breakpoint="mdAndUp"
    gutter={30}
    key="3"
    mobileBase={1}
  >
    <div />
  </styled.div>
  <styled.div
    base={3}
    breakpoint="mdAndUp"
    gutter={30}
    key="4"
    mobileBase={1}
  >
    <div />
  </styled.div>
  <styled.div
    base={3}
    breakpoint="mdAndUp"
    gutter={30}
    key="0"
    mobileBase={1}
  />
</div>
`;

exports[`List modifies the basis to prevent trailing items. 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-shrink: 0;
  -ms-flex-shrink: 0;
  flex-shrink: 0;
  -webkit-flex-grow: 0;
  -ms-flex-grow: 0;
  flex-grow: 0;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: bottom;
  -webkit-box-align: bottom;
  -ms-flex-align: bottom;
  align-items: bottom;
}

<div
  className="c0"
>
  <styled.div
    base={2}
    breakpoint="mdAndUp"
    gutter={30}
    key="0"
    mobileBase={1}
  >
    <div />
  </styled.div>
  <styled.div
    base={2}
    breakpoint="mdAndUp"
    gutter={30}
    key="1"
    mobileBase={1}
  >
    <div />
  </styled.div>
  <styled.div
    base={2}
    breakpoint="mdAndUp"
    gutter={30}
    key="2"
    mobileBase={1}
  >
    <div />
  </styled.div>
  <styled.div
    base={2}
    breakpoint="mdAndUp"
    gutter={30}
    key="3"
    mobileBase={1}
  >
    <div />
  </styled.div>
</div>
`;

exports[`List renders with a base and default mobileBase 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-shrink: 0;
  -ms-flex-shrink: 0;
  flex-shrink: 0;
  -webkit-flex-grow: 0;
  -ms-flex-grow: 0;
  flex-grow: 0;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: bottom;
  -webkit-box-align: bottom;
  -ms-flex-align: bottom;
  align-items: bottom;
}

<div
  className="c0"
>
  <styled.div
    base={2}
    breakpoint="mdAndUp"
    gutter={30}
    key="0"
    mobileBase={1}
  >
    <div />
  </styled.div>
  <styled.div
    base={2}
    breakpoint="mdAndUp"
    gutter={30}
    key="1"
    mobileBase={1}
  >
    <div />
  </styled.div>
</div>
`;

exports[`List renders with custom breakpoint 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-shrink: 0;
  -ms-flex-shrink: 0;
  flex-shrink: 0;
  -webkit-flex-grow: 0;
  -ms-flex-grow: 0;
  flex-grow: 0;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: bottom;
  -webkit-box-align: bottom;
  -ms-flex-align: bottom;
  align-items: bottom;
}

<div
  className="c0"
>
  <styled.div
    base={2}
    breakpoint="smAndUp"
    gutter={30}
    key="0"
    mobileBase={1}
  >
    <div />
  </styled.div>
  <styled.div
    base={2}
    breakpoint="smAndUp"
    gutter={30}
    key="1"
    mobileBase={1}
  >
    <div />
  </styled.div>
</div>
`;
